### PR TITLE
[4.2] Menu item custom field

### PR DESCRIPTION
**Pull request by brianteeman**
### Summary of Changes
adds a custom field to show a list of all menu items that can be selected and then displayed as a link in the frontend

### Testing Instructions
Install a prebuilt package
or
Install this PR, discover the extension and publish the plugin

Create a menuitem custom field. Check that it shows all menuitems from all menus in the select ;ist and then check on the front end that the link is displayed.

### Expected result AFTER applying this Pull Request
image

image

image

### Documentation Changes Required
New plugin

### CODE REVIEW

File: administrator/components/com_admin/sql/updates/mysql/4.2.0-2022-07-07.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `locked`, `manifest_cache`, `params`, `custom_data`, `ordering`, `state`) VALUES
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/mysql/4.2.0-2022-07-07.sql:1]: The „locked“ column should have value 1 and not 0.

File: administrator/components/com_admin/sql/updates/postgresql/4.2.0-2022-07-07.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder", "client_id", "enabled", "access", "protected", "locked", "manifest_cache", "params", "custom_data", "ordering", "state") VALUES
+(0, 'plg_fields_menuitem', 'plugin', 'menuitem', 'fields', 0, 1, 1, 0, 0, '', '', '', 0. 0);
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/postgresql/4.2.0-2022-07-07.sql]: The „locked“ column should have value 1 and not 0.

File: installation/sql/mysql/base.sql [hunk 3]
```diff
@@ -286,14 +286,15 @@ INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`,
...
 (0, 'plg_fields_media', 'plugin', 'media', 'fields', 0, 1, 1, 0, 1, '', '', '', 8, 0),
-(0, 'plg_fields_radio', 'plugin', 'radio', 'fields', 0, 1, 1, 0, 1, '', '', '', 9, 0),
-(0, 'plg_fields_sql', 'plugin', 'sql', 'fields', 0, 1, 1, 0, 1, '', '', '', 10, 0),
-(0, 'plg_fields_subform', 'plugin', 'subform', 'fields', 0, 1, 1, 0, 1, '', '', '', 11, 0),
-(0, 'plg_fields_text', 'plugin', 'text', 'fields', 0, 1, 1, 0, 1, '', '', '', 12, 0),
-(0, 'plg_fields_textarea', 'plugin', 'textarea', 'fields', 0, 1, 1, 0, 1, '', '', '', 13, 0),
-(0, 'plg_fields_url', 'plugin', 'url', 'fields', 0, 1, 1, 0, 1, '', '', '', 14, 0),
-(0, 'plg_fields_user', 'plugin', 'user', 'fields', 0, 1, 1, 0, 1, '', '', '', 15, 0),
-(0, 'plg_fields_usergrouplist', 'plugin', 'usergrouplist', 'fields', 0, 1, 1, 0, 1, '', '', '', 16, 0),
+(0, 'plg_fields_menuitem', 'plugin', 'menuitem', 'fields', 0, 1, 1, 0, '', '', '', 9, 0),
```
Comment (richard67) [inline installation/sql/mysql/base.sql]: The value for the “locked” column is missing here.

File: installation/sql/postgresql/base.sql [hunk 4]
```diff
@@ -292,14 +292,15 @@ INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder",
...
 (0, 'plg_fields_media', 'plugin', 'media', 'fields', 0, 1, 1, 0, 1, '', '', '', 8, 0),
-(0, 'plg_fields_radio', 'plugin', 'radio', 'fields', 0, 1, 1, 0, 1, '', '', '', 9, 0),
-(0, 'plg_fields_sql', 'plugin', 'sql', 'fields', 0, 1, 1, 0, 1, '', '', '', 10, 0),
-(0, 'plg_fields_subform', 'plugin', 'subform', 'fields', 0, 1, 1, 0, 1, '', '', '', 11, 0),
-(0, 'plg_fields_text', 'plugin', 'text', 'fields', 0, 1, 1, 0, 1, '', '', '', 12, 0),
-(0, 'plg_fields_textarea', 'plugin', 'textarea', 'fields', 0, 1, 1, 0, 1, '', '', '', 13, 0),
-(0, 'plg_fields_url', 'plugin', 'url', 'fields', 0, 1, 1, 0, 1, '', '', '', 14, 0),
-(0, 'plg_fields_user', 'plugin', 'user', 'fields', 0, 1, 1, 0, 1, '', '', '', 15, 0),
-(0, 'plg_fields_usergrouplist', 'plugin', 'usergrouplist', 'fields', 0, 1, 1, 0, 1, '', '', '', 16, 0),
+(0, 'plg_fields_menuitem', 'plugin', 'menuitem', 'fields', 0, 1, 1, 0, '', '', '', 9, 0),
```
Comment (richard67) [inline installation/sql/postgresql/base.sql]: The value for the “locked” column is missing here.

File: installation/sql/mysql/base.sql (hunk 3 above)
Comment (richard67) [reply installation/sql/mysql/base.sql]: This causes an SQL error at installation, so this has definitely not been tested by thy author.

File: installation/sql/mysql/base.sql [hunk 5]
```diff
@@ -286,14 +286,15 @@ INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`,
...
 (0, 'plg_fields_media', 'plugin', 'media', 'fields', 0, 1, 1, 0, 1, '', '', '', 8, 0),
-(0, 'plg_fields_radio', 'plugin', 'radio', 'fields', 0, 1, 1, 0, 1, '', '', '', 9, 0),
-(0, 'plg_fields_sql', 'plugin', 'sql', 'fields', 0, 1, 1, 0, 1, '', '', '', 10, 0),
-(0, 'plg_fields_subform', 'plugin', 'subform', 'fields', 0, 1, 1, 0, 1, '', '', '', 11, 0),
-(0, 'plg_fields_text', 'plugin', 'text', 'fields', 0, 1, 1, 0, 1, '', '', '', 12, 0),
-(0, 'plg_fields_textarea', 'plugin', 'textarea', 'fields', 0, 1, 1, 0, 1, '', '', '', 13, 0),
-(0, 'plg_fields_url', 'plugin', 'url', 'fields', 0, 1, 1, 0, 1, '', '', '', 14, 0),
-(0, 'plg_fields_user', 'plugin', 'user', 'fields', 0, 1, 1, 0, 1, '', '', '', 15, 0),
-(0, 'plg_fields_usergrouplist', 'plugin', 'usergrouplist', 'fields', 0, 1, 1, 0, 1, '', '', '', 16, 0),
+(0, 'plg_fields_menuitem', 'plugin', 'menuitem', 'fields', 0, 1, 1, 0, '1', '', '', 9, 0),
```
Comment (richard67) [inline installation/sql/mysql/base.sql]:
```suggestion
(0, 'plg_fields_menuitem', 'plugin', 'menuitem', 'fields', 0, 1, 1, 0, 1, '', '', '', 9, 0),
```
Still not right.

File: installation/sql/postgresql/base.sql [hunk 6]
```diff
@@ -292,14 +292,15 @@ INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder",
...
 (0, 'plg_fields_media', 'plugin', 'media', 'fields', 0, 1, 1, 0, 1, '', '', '', 8, 0),
-(0, 'plg_fields_radio', 'plugin', 'radio', 'fields', 0, 1, 1, 0, 1, '', '', '', 9, 0),
-(0, 'plg_fields_sql', 'plugin', 'sql', 'fields', 0, 1, 1, 0, 1, '', '', '', 10, 0),
-(0, 'plg_fields_subform', 'plugin', 'subform', 'fields', 0, 1, 1, 0, 1, '', '', '', 11, 0),
-(0, 'plg_fields_text', 'plugin', 'text', 'fields', 0, 1, 1, 0, 1, '', '', '', 12, 0),
-(0, 'plg_fields_textarea', 'plugin', 'textarea', 'fields', 0, 1, 1, 0, 1, '', '', '', 13, 0),
-(0, 'plg_fields_url', 'plugin', 'url', 'fields', 0, 1, 1, 0, 1, '', '', '', 14, 0),
-(0, 'plg_fields_user', 'plugin', 'user', 'fields', 0, 1, 1, 0, 1, '', '', '', 15, 0),
-(0, 'plg_fields_usergrouplist', 'plugin', 'usergrouplist', 'fields', 0, 1, 1, 0, 1, '', '', '', 16, 0),
+(0, 'plg_fields_menuitem', 'plugin', 'menuitem', 'fields', 0, 1, 1, 1, '', '', '', 9, 0),
```
Comment (richard67) [inline installation/sql/postgresql/base.sql]:
```suggestion
(0, 'plg_fields_menuitem', 'plugin', 'menuitem', 'fields', 0, 1, 1, 0, 1, '', '', '', 9, 0),
```

**Comment by woluweb** [comment]
I have tested this item :white_check_mark: successfully on b304e06e56edc780697289ba42371ae7b8c94e4f<br /><br />Hi,

Txs very much Brian for introducing this new Type of Custom Fields!

Note for those who don't know: this follows a question I raised yesterday on the Joomla Facebook Group: https://www.facebook.com/groups/joomlanospam/posts/10158401801610997/

I have tested this successfully (on an existing J4.1 website).
You can see the result on https://subform.joomlacustomfields.org/ (link "Matrioshka" at the bottom of Article 1).

This new Type of Custom Fields can be useful in many cases.
But let me share here my own usecase as it could give ideas to others: suppose you have a Blog View... but when clicking on one of the post you don't want to open the corresponding article... but *another* Blog View. With other words, you want "**nested Blog Views**" (ie a kind of "Matryoshka").

Txs again Brian for this (very) quick PR :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38244">issues.joomla.org/tracker/joomla-cms/38244</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on b304e06e56edc780697289ba42371ae7b8c94e4f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38244">issues.joomla.org/tracker/joomla-cms/38244</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38244">issues.joomla.org/tracker/joomla-cms/38244</a>.</sub>

**Comment by roland-d** [comment]
Thank you.

**Comment by HLeithner** [comment]
@chmst found out that this field breaks if you set a default value
image

**Comment by roland-d** [comment]
So I checked and the `Default value` has no use in this field. However there is no toggle we can set to not show it for a specific field. Due to being there and when filled, it is added as XML element to the list of options to render and this crashes because it is not used and as such does not match anything and the default is throw this error message.

Since the default makes no sense here, I will create a PR where the field itself can decide if the `Default value` should be shown.

**Comment by roland-d** [comment]
PR created: #38423

**Comment by roland-d** [comment]
This PR has unfortunately been reverted as there are more issues to be resolved. Related PRs with fixes are:
- #38426
- #38457

 It is best to roll this PR and the other PRs into 1 new PR for 4.3, so we have more time to test and verify the solutions.

**Comment by brianteeman** [comment]
only because we both misread each others comments

**Comment by laoneo** [comment]
Best thing would be to offer the plugin in the JED and not bloat the core.

**Comment by brianteeman** [comment]
Why? Then we should remove all the plugins that are for core fields. Honestly that makes zero sense.

**Comment by laoneo** [comment]
Because core has already over 200 extensions when you do a blank installation, which I think is more than enough. Instead of, we should try to push the extension directory.

**Pull request merged**